config/moved_config.go

```diff
@@ -97,6 +100,31 @@ func (cfg *Config) Validate() error {
 }
 
 func (cfg *Config) validateService() error {
+
+	zapCfg := &zap.Config{
+		Level:       zap.NewAtomicLevelAt(zapcore.InfoLevel),
+		Development: cfg.Telemetry.Logs.Development,
+		Sampling: &zap.SamplingConfig{
+			Initial:    100,
+			Thereafter: 100,
+		},
+		Encoding:          cfg.Telemetry.Logs.Encoding,
+		EncoderConfig:     zap.NewProductionEncoderConfig(),
+		OutputPaths:       cfg.Telemetry.Logs.OutputPaths,
+		ErrorOutputPaths:  cfg.Telemetry.Logs.ErrorOutputPaths,
+		DisableCaller:     cfg.Telemetry.Logs.DisableCaller,
+		DisableStacktrace: cfg.Telemetry.Logs.DisableStacktrace,
+		InitialFields:     cfg.Telemetry.Logs.InitialFields,
+	}
+	if zapCfg.Encoding == "console" {
+		// Human-readable timestamps for console format of logs.
+		zapCfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+	}
+	logger, err := zapCfg.Build()
+	if err != nil {
+		return err
+	}
```

Comment:
Why this change?

Reply:
> > Oh, I see the confusion now. The example shows how one would proceed in changing the signature of a function to add a new return value (an error).
> > In your case, you should just add the new condition to the function. But instead of returning an error, you should just log a warning for a couple of releases and then return an error. In this case, given the bug is because the validation does not exist for this condition, adding the condition wouldn't be considered a breaking change (even though it would indeed break already faulty instances).
> > cc @open-telemetry/collector-contrib-approvers, for more opinions on this.
> 
> Ok, I got it. So I need to do following things
> 
> * remove `mustValidateService` function, just add new validation to original `validateService` function
> * for next release(N+1), just log a warning without return error, so the instance with faulty config can still start.
> * for N+2 release, return error for this validation.
> 


Becase I talked about the with @jpkrohling , we should first log with warning for next release(N+1), and return error in N+2 release. So I have to output a warning log in validateService function of config/move_config.go. 

But for now, the config.Config does not have a logger field in it. In order to keep the same logging format as collector, so I initialize a logger object same as https://github.com/open-telemetry/opentelemetry-collector/blob/e98116922cf6c079c246fa78eef59dfd9c2e4867/service/internal/telemetrylogs/logger.go#L26

I also think this is not the right way to output warning log in config/move_config.go, do you have any suggestion for this? @bogdandrutu

Reply:
It does; would that be a bit too much though? Not sure I see users wanting to enable this feature gate themselves.

Reply:
Honestly, I prefer the logging way, since the validation of config may be enhanced again in the future, not only in this pull request, warning log may be more friendly to both developers and users.

@bogdandrutu @jpkrohling

Reply:
Users will start to ask to configure that, etc. If you want to log something then simply return an error with a good message. Also why configuring a `zap.Logger` vs `fmt.Printf`?

Reply:
Honestly, how to log the warning message also puzzled me for a while. Finally I choose `zap.Logger` because I want to keep the same logging experience with other logs for collector, as following (the first line):

```
2022-09-29T09:17:18.442+0800	warn	config/moved_config.go:180	Telemetry config validation failed	{"reason": "collector telemetry metric address should exist when metric level is not none"}

2022-09-29T09:17:18.442+0800	info	service/telemetry.go:107	Skipping telemetry setup.	{"address": "", "level": "basic"}
2022-09-29T09:17:18.442+0800	info	components/components.go:30	In development component. May change in the future.	{"kind": "exporter", "data_type": "traces", "name": "logging", "stability": "in development"}
```

If use `fmt.Printf` instead, the experience will be
```
Telemetry config validation failed: collector telemetry metric address should exist when metric level is not none

2022-09-29T09:22:57.747+0800	info	service/telemetry.go:107	Skipping telemetry setup.	{"address": "", "level": "basic"}
2022-09-29T09:22:57.747+0800	info	components/components.go:30	In development component. May change in the future.	{"kind": "exporter", "data_type": "traces", "name": "logging", "stability": "in development"}
```

I'm not saying keep the uniform logging experience is necessary, I think we can talk more about this and agree on a right way.

@bogdandrutu

Reply:
Do you have more suggestions? @bogdandrutu

Reply:
ping @bogdandrutu @jpkrohling 

Do you have any suggestion for push forwarding this pr?

Reply:
I still think that we should log a warning message in case we have an empty telemetry metrics address and level is not none, returning an error in a future version. I don't think this deserves to be behind a feature gate, though. But I'm good with whatever @bogdandrutu decides.

Reply:
> Honestly, how to log the warning message also puzzled me for a while. Finally I choose `zap.Logger` because I want to keep the same logging experience with other logs for collector, as following (the first line):
> 
> ```
> 2022-09-29T09:17:18.442+0800	warn	config/moved_config.go:180	Telemetry config validation failed	{"reason": "collector telemetry metric address should exist when metric level is not none"}
> 
> 2022-09-29T09:17:18.442+0800	info	service/telemetry.go:107	Skipping telemetry setup.	{"address": "", "level": "basic"}
> 2022-09-29T09:17:18.442+0800	info	components/components.go:30	In development component. May change in the future.	{"kind": "exporter", "data_type": "traces", "name": "logging", "stability": "in development"}
> ```
> 
> If use `fmt.Printf` instead, the experience will be
> 
> ```
> Telemetry config validation failed: collector telemetry metric address should exist when metric level is not none
> 
> 2022-09-29T09:22:57.747+0800	info	service/telemetry.go:107	Skipping telemetry setup.	{"address": "", "level": "basic"}
> 2022-09-29T09:22:57.747+0800	info	components/components.go:30	In development component. May change in the future.	{"kind": "exporter", "data_type": "traces", "name": "logging", "stability": "in development"}
> ```
> 
> I'm not saying keep the uniform logging experience is necessary, I think we can talk more about this and agree on a right way.
> 
> @bogdandrutu

@bogdandrutu Do you have more suggestions? 
I think we can all agree with the opinion that we should use warning logs, not feature-gates.
Since this package does not have import `zap.Logger` , should we just use fmt.Printf?

Reply:
Yes, let's use `fmt.Printf` then

Reply:
Could you help review this again? @bogdandrutu

Reply:
Ping @bogdandrutu